### PR TITLE
[PnP] Adding Properties E2E Tests

### PR DESCRIPTION
**Pull request by jamdavi**
This PR adds end-to-end tests for properties operations.

### CODE REVIEW

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+            var propName = Guid.NewGuid().ToString();
+
+            logger.Trace($"{nameof(Properties_DeviceSetsPropertyAndGetsItBackAsync)}: name={propName}, value={propValue}");
+
+            var props = new ClientPropertyCollection();
+            props[propName] = propValue;
+            await deviceClient.UpdateClientPropertiesAsync(props).ConfigureAwait(false);
+
+            // Validate the updated twin from the device-client
+            Twin deviceTwin = await deviceClient.GetTwinAsync().ConfigureAwait(false);
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:219]: GetClientPropertiesAsync instead of GetTwinAsync?

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+                        [propName1] = new Dictionary<string, object>
+                        {
+                            [propName2] = null
+                        }
+                    })
+                .ConfigureAwait(false);
+            serviceTwin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+            Assert.IsTrue(serviceTwin.Properties.Reported.Contains(propName1));
+            string value2 = serviceTwin.Properties.Reported[propName1].ToString();
+            Assert.AreEqual(value2, propEmptyValue);
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:449]: repeated?

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+            await deviceClient.UpdateClientPropertiesAsync(patch).ConfigureAwait(false);
+            await deviceClient.CloseAsync().ConfigureAwait(false);
+
+            Twin serviceTwin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+            Assert.AreEqual<string>(serviceTwin.Properties.Reported[propName].ToString(), propValue);
+
+            Logger.Trace("verified " + serviceTwin.Properties.Reported[propName].ToString() + "=" + propValue);
+        }
+
+        private async Task Properties_ServiceDoesNotCreateNullPropertyInCollectionAsync(Client.TransportType transport)
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:406]: how about we also we add a test that verifies that updating an existing property patch with null value removes the property from the twin

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+            Assert.IsFalse(serviceTwin.Properties.Reported.Contains(propName1));
+
+            await deviceClient
+                .UpdateClientPropertiesAsync(
+                    new ClientPropertyCollection
+                    {
+                        [propName1] = new Dictionary<string, object>
+                        {
+                            [propName2] = null
+                        }
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:433]: interesting, I wasn't aware that propName2 wouldn't make it to the final twin

Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:433]: I thought the removal of "null" value properties referred to only root level properties, and not nested ones.

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+            }
+            catch (IotHubException)
+            {
+                exceptionThrown = true;
+            }
+
+            Assert.IsTrue(exceptionThrown, "IotHubException was expected for updating reported property with an invalid property name, but was not thrown.");
+
+            Twin serviceTwin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+            Assert.IsFalse(serviceTwin.Properties.Reported.Contains(propName1));
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:472]: does propName2 make it through? is it an all-or-none operation? we should add explicit assertions for propeName2 as well, whatever be the case.

File: e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+            {
+                Assert.AreEqual(JsonConvert.SerializeObject(propFromCollection), JsonConvert.SerializeObject(propValue));
+            }
+            else
+            {
+                Assert.Fail($"The property {propName} was not found in the  collection");
+            }
+            // Validate the updated twin from the service-client
+            Twin completeTwin = await _registryManager.GetTwinAsync(deviceId).ConfigureAwait(false);
+            var actualProp = completeTwin.Properties.Reported[ComponentName][propName];
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs:234]: nit: dynamic instead of var?

File: e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+            using var registryManager = RegistryManager.CreateFromConnectionString(Configuration.IoTHub.ConnectionString);
+
+            var twinPatch = new Twin();
+            twinPatch.Properties.Desired[componentName] = new
+            {
+                __t = "c",
+            };
+            if (propValue is List<object>)
+            {
+                twinPatch.Properties.Desired[componentName][propName] = (Newtonsoft.Json.Linq.JToken)(JsonConvert.DeserializeObject(JsonConvert.SerializeObject(propValue)));
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs:290]: why do we need to serialize and deserialize it a list value? isn't Twin able to accept the object as-is?

File: e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+
+            await deviceClient
+                .UpdateClientPropertiesAsync(
+                    new ClientPropertyCollection
+                    {
+                        { ComponentName, propName1, null }
+                    })
+                .ConfigureAwait(false);
+            Twin serviceTwin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+            Assert.IsTrue(serviceTwin.Properties.Reported.Contains(ComponentName));
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs:434]: assertion for component name containing an empty value? we should probably check with service if that is the expected behavior, i.e. to remove an updated property we pass in an empty value, how to remove a component?

File: e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+            }
+            catch (IotHubException)
+            {
+                exceptionThrown = true;
+            }
+
+            Assert.IsTrue(exceptionThrown, "IotHubException was expected for updating reported property with an invalid property name, but was not thrown.");
+
+            Twin serviceTwin = await registryManager.GetTwinAsync(testDevice.Id).ConfigureAwait(false);
+            Assert.IsFalse(serviceTwin.Properties.Reported.Contains(propName1));
```
Comment (abhipsaMisra) [inline e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs:491]: shouldn't this assertion be `serviceTwin.Properties.Reported[ComponentName].Contains(propName1)`?

**Review by abhipsaMisra — APPROVED** [review]
A couple of questions

**Comment by abhipsaMisra** [comment]
I wonder if it is adding fault injection tests for the two new APIs (GetClientProperties, UpdateClientProperties), since the transport layer is new for those.

File: e2e/test/iothub/properties/PropertiesE2ETests.cs (hunk 3 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:350]: added

File: e2e/test/iothub/properties/PropertiesE2ETests.cs (hunk 5 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:416]: it is all-or-non, none of the properties make it through

File: e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs (hunk 7 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs:288]: don't need this

File: e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs (hunk 8 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs:426]: done

File: e2e/test/Helpers/TestDeviceCallbackHandler.cs [hunk 10]
```diff
@@ -48,33 +51,42 @@ public Message ExpectedMessageSentByService
...
-                        _logger.Trace($"{nameof(SetDeviceReceiveMethodAsync)}: Error during DeviceClient callback method: {ex}.");
+            get => Volatile.Read(ref _expectedClientPropertyValue);
+            set => Volatile.Write(ref _expectedClientPropertyValue, value);
+        }
 
-                        _methodExceptionDispatch = ExceptionDispatchInfo.Capture(ex);
-                        return Task.FromResult(new MethodResponse(500));
-                    }
-                    finally
+        public async Task SetDeviceReceiveMethodAsync(string methodName, string deviceResponseJson, string expectedServiceRequestJson)
```
Comment (abhipsaMisra) [inline e2e/test/Helpers/TestDeviceCallbackHandler.cs:60]: this change is only fixing the indentation

File: e2e/test/Helpers/TestDeviceCallbackHandler.cs [hunk 11]
```diff
@@ -87,29 +99,32 @@ public async Task SetTwinPropertyUpdateCallbackHandlerAsync(string expectedPropN
...
-                    _logger.Trace($"{nameof(SetTwinPropertyUpdateCallbackHandlerAsync)}: DeviceClient {_testDevice.Id} callback twin: DesiredProperty: {patch}, {context}");
-
-                    try
-                    {
-                        string propertyValue = patch[expectedPropName];
-                        propertyValue.Should().Be(ExpectedTwinPropertyValue, "The property value should match what was set by service");
-                        context.Should().Be(userContext, "The context should match what was set by service");
-                    }
-                    catch (Exception ex)
+            await _deviceClient
```
Comment (abhipsaMisra) [inline e2e/test/Helpers/TestDeviceCallbackHandler.cs:102]: this change is only fixing the indentation

File: e2e/test/Helpers/TestDeviceCallbackHandler.cs [hunk 12]
```diff
@@ -120,31 +135,33 @@ public async Task WaitForTwinCallbackAsync(CancellationToken ct)
 
         public async Task SetMessageReceiveCallbackHandlerAsync()
         {
-            await _deviceClient.SetReceiveMessageHandlerAsync(
-                async (receivedMessage, context) =>
-                {
-                    _logger.Trace($"{nameof(SetMessageReceiveCallbackHandlerAsync)}: DeviceClient {_testDevice.Id} received message with Id: {receivedMessage.MessageId}.");
-
-                    try
+            await _deviceClient
```
Comment (abhipsaMisra) [inline e2e/test/Helpers/TestDeviceCallbackHandler.cs:138]: this change is only fixing the indentation

File: e2e/test/Helpers/TestDeviceCallbackHandler.cs [hunk 13]
```diff
@@ -158,6 +175,48 @@ public async Task WaitForReceiveMessageCallbackAsync(CancellationToken ct)
             _receiveMessageExceptionDispatch?.Throw();
         }
 
+        public async Task SetClientPropertyUpdateCallbackHandlerAsync<T>(string expectedPropName, string componentName = default)
```
Comment (abhipsaMisra) [inline e2e/test/Helpers/TestDeviceCallbackHandler.cs:178]: new callback handler for property update notifications

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+
+        private static readonly Dictionary<string, object> s_mapOfPropertyValues = new Dictionary<string, object>
+        {
+            { "key1", 123 },
+            { "key2", "someString" },
+            { "key3", true }
+        };
+
+        [LoggedTestMethod]
+        public async Task Properties_DeviceSetsPropertyAndGetsItBack_Mqtt()
```
Comment (vinagesh) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:36]: I know we follow this style of tests in our code. But for the future something to consider for better readability of e2e tests is to add sections for `arrange` `act` `assert` and use [Theory] with multiple protocol inputs. It is very hard to understand what is being setup, what is being called and checked for like this. There might be some amount of repeated code with the arrange, act, assert format but each test will be very readable.

File: iothub/device/src/PayloadCollection.cs [hunk 15]
```diff
@@ -33,6 +33,14 @@ public abstract class PayloadCollection : IEnumerable<object>
         /// <remarks>
         /// This accessor is best used to access and cast to simple types.
         /// It is recommended to use <see cref="TryGetValue"/> to deserialize to a complex type.
+        /// <para>
+        /// <remarks>
+        /// For setting component-level property values see <see cref="ClientPropertyCollection.AddComponentProperty(string, string, object)"/>
+        /// and <see cref="ClientPropertyCollection.AddComponentProperties(string, IDictionary{string, object})"/> instead.
+        /// These convenience methods ensure that component-level properties include the component identifier markers { "__t": "c" }.
+        /// For more information see <see href="https://docs.microsoft.com/en-us/azure/iot-pnp/concepts-convention#sample-multiple-components-read-only-property"/>.
```
Comment (barustum) [inline iothub/device/src/PayloadCollection.cs:41]: en-us should not be included in uri

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 16]
```diff
@@ -0,0 +1,448 @@
...
+
+            logger.Trace($"{nameof(Properties_DeviceSetsPropertyAndGetsItBackAsync)}: name={propName}, value={propValue}");
+
+            var props = new ClientPropertyCollection();
+            props.AddRootProperty(propName, propValue);
+            await deviceClient.UpdateClientPropertiesAsync(props).ConfigureAwait(false);
+
+            // Validate the updated properties from the device-client
+            ClientProperties clientProperties = await deviceClient.GetClientPropertiesAsync().ConfigureAwait(false);
+            bool isPropertyPresent = clientProperties.TryGetValue<T>(propName, out T propFromCollection);
```
Comment (vinagesh) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:213]: propValueFromCollection?

File: iothub/device/src/ClientPropertyCollection.cs [hunk 17]
```diff
@@ -297,18 +296,20 @@ internal static ClientPropertyCollection FromClientTwinDictionary(IDictionary<st
         /// <param name="componentName">The component with the properties to add or update.</param>
         /// <param name="forceUpdate">Forces the collection to use the Add or Update behavior.
         /// Setting to true will simply overwrite the value. Setting to false will use <see cref="IDictionary{TKey, TValue}.Add(TKey, TValue)"/></param>
-        /// <exception cref="ArgumentNullException"><paramref name="properties"/> is <c>null</c>. </exception>
+        /// <exception cref="ArgumentNullException"><paramref name="properties"/> is <c>null</c> for a root-level property operation. </exception>
```
Comment (barustum) [inline iothub/device/src/ClientPropertyCollection.cs:299]: space after '.'

File: e2e/test/iothub/properties/PropertiesE2ETests.cs (hunk 14 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:36]: I agree. I've found unit tests to be very easy to use the arrange-act-assert format, but since end-to-end tests usually have these actions spread out within the test, I've found those comments unnecessarily cluttering up the test. We can try it out in a couple of tests and see how we feel.

Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:36]: I haven't used `[Theory]` before, is that like parameterized tests?

File: iothub/device/src/PayloadCollection.cs (hunk 15 above)
Comment (abhipsaMisra) [reply iothub/device/src/PayloadCollection.cs:41]: good catch!

File: e2e/test/iothub/properties/PropertiesE2ETests.cs (hunk 16 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:213]: sure, I don't mind renaming this variable

File: e2e/test/iothub/properties/PropertiesE2ETests.cs [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+                await deviceClient
+                    .UpdateClientPropertiesAsync(
+                        new ClientPropertyCollection
+                        {
+                            [propName1] = 123,
+                            [propName2] = "abcd"
+                        })
+                    .ConfigureAwait(false);
+            };
+            await func.Should().ThrowAsync<IotHubException>();
```
Comment (vinagesh) [inline e2e/test/iothub/properties/PropertiesE2ETests.cs:433]: Does this throw a specific code for the exception that we can validate here based on protocol?

File: e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs [hunk 19]
```diff
@@ -0,0 +1,480 @@
...
+                                ComponentName,
+                                new Dictionary<string, object> {
+                                    { propName1, 123 },
+                                    { propName2, "abcd" }
+                                }
+                            }
+                        })
+                    .ConfigureAwait(false);
+            };
+            await func.Should().ThrowAsync<IotHubException>();
```
Comment (vinagesh) [inline e2e/test/iothub/properties/PropertiesWithComponentsE2ETests.cs:466]: same, any code that can be validated?

**Review by vinagesh — APPROVED** [review]

File: e2e/test/iothub/properties/PropertiesE2ETests.cs (hunk 18 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:433]: It does not, no. Since this is mqtt, we don't get the advantage of receiving descriptive error codes from the service.

File: e2e/test/iothub/properties/PropertiesE2ETests.cs (hunk 14 above)
Comment (vinagesh) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:36]: yupp those are parameterized tests. Theory is in xunit but it is called datarow in ms test Ex of something I added recently - https://github.com/Azure/azure-iot-sdk-csharp/blob/54e8280f775f7c6db00a29b98169588532180fc2/e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs#L45

I feel writing tests using 'arrange act assert' makes it very readable and is a commonly followed format.

Comment (abhipsaMisra) [reply e2e/test/iothub/properties/PropertiesE2ETests.cs:36]: Ah, I see. Parameterized MS tests don't support parallelization yet, so we have refrained from using those in our E2E tests suite since it'll end up increasing the test run time.
